pkg/test/impl/driver/instance.go

```diff
@@ -81,31 +75,36 @@ func (d *driver) Initialize(a *Args) error {
 	if err := args.Validate(); err != nil {
 		return err
 	}
-	d.args = args
 
-	d.suiteDependencies = args.SuiteDependencies
-
-	var env environment.Interface
+	// Initialize the environment.
+	var env internal.Environment
 	var err error
-
 	switch a.Environment {
 	case EnvLocal:
-		env, err = local.NewEnvironment(d)
-
+		env, err = local.NewEnvironment()
 	case EnvKubernetes:
-		env, err = cluster.NewEnvironment(d, a.KubeConfig)
-
+		env, err = cluster.NewEnvironment(a.KubeConfig)
 	default:
 		return fmt.Errorf("unrecognized environment: %s", a.Environment)
 	}
 
 	if err != nil {
 		return fmt.Errorf("unable to initialize environment '%s': %v", a.Environment, err)
 	}
-	d.env = env
 
-	d.testID = args.TestID
-	d.runID = generateRunID(args.TestID)
+	// Create the context, but do not attach it to the member fields before initializing.
+	ctx := internal.NewTestContext(
+		args.TestID,
+		generateRunID(args.TestID),
+		a.WorkDir,
+		env)
+
+	if err = env.Initialize(ctx); err != nil {
```

Comment:
Just out of curiosity, why are the environment and context separate? The context just seems to be a data structure and there is a 1-to-1 mapping between context and environment AFAICT. The same thing could be accomplished by giving all environments the properties TestID, RunID, and WorkDir.  But, I'm probably missing something.

Reply:
internal.Environment interface is to be implemented by environments to provide common, internal services to the main test driver. TestContext contains ambient test context data that is not necessarily part of the environment only. It is meant to be used internally by all parts of the test framework. It is more closely aligned with the driver, which I've tried using instead of context initially. But that creates a cyclic reference problem between environments and driver.

We can make environments implement some of this, or refactor the code in such a way that the dependency chain is broken. But my past experience is, generally having a single internal context object (that is somewhat of a grab-bag) actually helps simplify the other parts of the code, and let them only focus on the particulars they care about. 

In fact, I am currently not very happy with the current reality, as it still causes a significant amount of duplication between local and cluster environments. I'd like to reduce that as well. But I haven't tackled that yet.

Reply:
ok, it's fine for now.  I suspect much of this will evolve over time anyway.